nds-org/ndslabs
# Support real PVCs as a replacement GLFS + hostPath

# Problem
Fixes #249. Workbench currently relies on an admittedly fragile storage model.

Currently, we run GlusterFS in a set of server pods (even-numbered) on our storage nodes. On every node in the cluster we run a Gluster client pod as well. The client pod reaches out to the server pods to mount the Gluster volume onto the host machine from within the container.

The reasoning behind this approach was our requirement for ReadWriteMany volume support, and the lack for such support in our initial OpenStack environments. 

# Approach
Now that we have more thoroughly explored PVC support in OpenStack environments without making any assumptions about underlying infrastructure, our cup runneth over with different ways to support dynamically-provisioned ReadWriteMany storage using PVCs.

This new approach does away with the old hostPath / GlusterFS implementation and replaces it with `PersistentVolumeClaims`. Since many types of PVCs can support ReadWriteMany, this opens up several different storage backing options (depending on the scenario):
* `MinikubeStorageProvisioner`: Minikube instances come with the `storage-provisioner` addon enabled by default
* `HostPathProvisioner`: single-node development instances can still use PVCs by using a provisioner that simply performs a `mkdir` locally
* `NFSProvisioner`: small, multi-node clusters can use PVCs the same way, with a provisioner that simply performs a `mkdir` remotely 
* `GlusterfsSimpleProvisioner`: clusters that have heavy storage needs or scalability concerns can opt to use something heavier like GlusterFS - this allows you to add new storage as needed while using GlusterFS for logical volume management (e.g. multiple physical volumes appearing as one logical volume)

# How to test

1. Follow the steps here to get a development environment up and running: https://github.com/nds-org/ndslabs/tree/develop/apiserver
2. Check out this branch: `git fetch --all && git checkout pvc-support`
3. Edit `apiserver.json`
    * At a bare minimum, you'll need to set `kubernetes.address` / `specs.path` / `support.email` / `supportEmail` to point to the correct locations
    * You can leave `kubernetes.pvcStorageClass` blank for now to use the cluster default StorageClass, or specify your own StorageClass here
4. Rebuild and restart the API server:
```bash
./build.sh local
./build/bin/apiserver-darwin-amd64   # For OSX
./build/bin/apiserver-linux-amd64    # For Linux
```
5. Edit `postman/workbench.postman_environment.json` to set the correct values for (at least) the `host` and `email` field
```bash
vi postman/workbench.postman_environment.json
```
6. Use `newman` to run the tests:
```bash
npm install -g newman
cd postman/
newman run --insecure  --environment=workbench.postman_environment.json --delay-request=1000 Workbench.postman_collection_local.json
```
7. Open a new terminal window and execute `kubectl get pvc ,n test --watch`
    * You should see a new line item when PVCs are created or change state
8. Run `pvc-explorer.yaml` to easily examine the contents of your PVC:
```bash
echo 'apiVersion: v1
kind: ReplicationController
metadata:
  name: pvc-explorer
  namespace: test
spec:
  replicas: 1
  selector:
    component: pvc-explorer
  template:
    metadata:
      labels:
        component: pvc-explorer
    spec:
      hostNetwork: true
      containers:
      - name: pvc-explorer
        imagePullPolicy: Always
        image: ubuntu:16.04
        stdin: true
        tty: true
        workingDir: "/pvc/"
        volumeMounts:
        - name: pvc
          mountPath: "/pvc/"
      volumes:
      - name: pvc
        persistentVolumeClaim:
          claimName: test-home' > pvc-explorer.yaml && kubectl create -f pvc-explorer.yaml
```
    * Performing an `ls -al`, you should see an AppData subdirectory. Drilling down into this directory, you should see a familiar MongoDB file structure
